liquidator/liquidator.js

```diff
@@ -285,7 +287,29 @@ class Liquidator {
       // Send the transaction or report failure.
       let receipt;
       try {
-        receipt = await liquidation.send(txnConfig);
+        // Make sure to keep trying with this nonce
+        const nonce = await this.web3.eth.getTransactionCount(this.account);
+
+        // Min Gas Price, with a max gasPrice of double
+        const minGasPrice = parseInt(this.gasEstimator.getCurrentFastPrice(), 10);
+        const maxGasPrice = 2 * minGasPrice;
+
```

Comment:
As a side comment would it be correct in saying that this is "blocking" logic? in the sense that that the bot will effectively wait at this point in the code (within this try/catch) until the tx is included within a block? 

This means that the bot is basically only capable of processing 1 liquidation at a time. In the worst-case scenario, this means that if there are _many_ cascading liquidations we can only process one every ~5 mins (based on the each tx taking ~`liquidationDeadline` to be included). Is this something we would want to consider doing differently?

I would say having a more aggressive `gasPriceScalingFunction` would hopefully ensure that txs are included much faster.

Thoughts on this consideration?

Reply:
Yeap! That is correct. But isn't this the wanted behavior?

[According to the web3 docs](https://web3js.readthedocs.io/en/v1.2.7/callbacks-promises-events.html) the `awaited` promise will only yield once the transaction receipt is returned (which only happens once the transaction is mined).

We can definitely have a more aggreesive `gasPriceScalingFunction`. There is also an `EXPONENTIAL` function if that pleases you :stuck_out_tongue:

Reply:
Actually [after reading this discussion thread](https://github.com/ethereum/web3.js/issues/1547) I'm not too sure anymore. Couldn't you just not await `queryAndLiquidate` in that case if you'd like it to be non-blocking?

Reply:
ye, I dont think removing the `await` is a good solution as we would basically "loose" the promise then. This is not ideal though because we really need to _track_ the tx until it is within a block. 

Otherwise, I could foresee creating a transaction handler that responsible for this and runs in a separate thread from the main bot logic. In this case we would basically give the tx handler the transaction and it would watch it, bump gas according to what ynatm tells it and the rest of the bot logic can keep running in a non-blocking way. This would be more complex but would also remove the blocking behavior we have now. WDYT?

Reply:
@kendricktan?

Reply:
As JS is an single threaded event driven language, I fear that adding worker threads will over complicate this.

If you would like to make it non-blocking, perhaps the best way would be to do it via promise chaining without the `async/await` keywords

But isn't the existing logic for a lot of the bots blocking either way? e.g.

```javascript
await liquidation.send({...})
```